tree: clean up DOidWrapper usage

yuzefovich opened the pull request:
This PR cleans up usage of `DOidWrapper` utility. Throughout the years we've accumulated some dead code as well as in many cases we had blind copy-paste of unnecessary code, and this PR gets rid of it. Namely, `DOidWrapper` utility is now used to support `DName` and `DRefCursor` (on top of `DString`) and `DCIText` (on top of `DCollatedString`), and any other usages are forbidden. This PR removes all those other usages.

Please see each commit for more details.

Epic: None

yuzefovich committed:
tree: remove DOidWrapper support for DArrays

Long time ago we implemented `int2vector` and `oidvector` types using
DOidWrapper around the arrays. Some time later, but also long time ago
in 3f4d1a09d70770380ecce6d755124f4388282469 we removed those usages by
pushing the "custom" Oid directly into DArray datum, which made the code
for supporting DOidWrapper for arrays redundant. This commit removes
that dead code as well as `AsDArray` function in favor of explicit type
cast.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.svg" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/150454)
<!-- Reviewable:end -->

yuzefovich committed:
tree: remove DOidWrapper support for DInt

Initially, in 9c9b045bc819e9ef08a0af1f56bde70ffbc16b8d, when we added
DOidWrapper thingy in the first place, we supported Oid type using it.
Soon afterwards in d2526b39111339e49cf0c294a6b16d32ba65ce5e we
introduced separate DOid datum to support the Oid type, so we removed
that usage of DOidWrapper, which made the relevant code redundant. This
commit removes it.

Release note: None

yuzefovich committed:
sql: remove redundant DOidWrapper checks

We had many `tree.AsD*` functions which checked whether the given datum
was a DOidWrapper around another datum. Given that we now only support
oid-wrapping on top of `DString` and `DCollatedString`, most of these
functions are redundant and were only introduced due to blind
copy-paste. This commit removes most of them, replacing with an explicit
type cast.

Release note: None

yuzefovich committed:
tree: audit blind assertions to *DCollatedString

Now that we support CIText via the DOidWrapper on top of
DCollatedString, we need to be careful when using "blind" type
assertions like `cs := d.(*tree.DCollatedString)` since it might fail.
This commit audits all of these assertions in non-test code and fixes
them to use `AsDCollatedString` where appropriate.

Release note: None

mgartner reviewed (approved):
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.io/lgtm.png" height="20" width="61"/> Nice cleanup! Thank you!

> @mgartner reviewed 6 of 6 files at r1, 16 of 16 files at r2, 3 of 3 files at r3, 17 of 17 files at r4, all commit messages.
> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/150454#-:-O_D0G-X9tmhPHnv0B_q:b-popyyh)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @DrewKimball)



<!-- Sent from Reviewable.io -->

yuzefovich commented:
TFTR!

bors r+
